src/utils/gcsApi.ts

```diff
@@ -0,0 +1,321 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import {
+  Bucket,
+  Storage,
+  UploadOptions as GCSUploadOptions,
+} from '@google-cloud/storage';
+import * as ConsolaRaw from 'consola';
+import { isDryRun } from 'dryrun';
+
+import { logger as loggerRaw } from '../logger';
+import { reportError } from './errors';
+import { checkEnvForPrerequisite, RequiredConfigVar } from './env';
+
+let logger: typeof ConsolaRaw;
+
+const DEFAULT_MAX_RETRIES = 5;
+const DEFAULT_UPLOAD_METADATA = { cacheControl: `public, max-age=300` };
+
+const IS_DRY_RUN = isDryRun();
+
+/**
+ * Mapping between file extension regexps and the corresponding content type
+ * that will be set.
+ */
+const CONTENT_TYPES_EXT: Array<[RegExp, string]> = [
+  [/\.js$/, 'application/javascript; charset=utf-8'],
+  [/\.js\.map$/, 'application/json; charset=utf-8'],
+];
+
+/**
+ * Is this bucket being used as an artifact store or a target?
+ */
+export const enum BucketRole {
```

Comment:
Yeah, it's a valid point - should the utility class be aware of how it's being used or not? I did that in order to enable it to behave differently depending on what it was being used for.

In the end, the only place where that came into play was tagging the logger appropriately (in the constructor), which is why I don't actually store the value past that point. (Which answers your second question as well - it's in the config object because that's what gets passed to the constructor.)

I guess I'm not entirely sure how this is different from something like the credentials, which are also in the config and also discarded as soon as we've set up the bucket.

Reply:
Oh, wait - scratch that. Setting the logger tag is not the only place `bucketRole` is used (I gave it a different name for some reason, which is why my search didn't find it; fixed now). It's also used to know which env variables to look for (in `getGCSCredsFromEnv`), and I'm not immediately seeing a way around that.

Reply:
> I guess I'm not entirely sure how this is different from something like the credentials, which are also in the config and also discarded as soon as we've set up the bucket.

I think what bothers me here is that the user of the function doesn't really care about that logger prefix, but we require it as part of `GCSBucketConfig`.
I don't mind keeping the enum for picking up the right env variables, but not sure that the logger prefix is worth it.

Reply:
> Let's also make this a "regular" function :)

Okay, so to confirm:

- Move `getGCSCredsFromEnv` to be a function in the module (rather than a static method on the class)
- Pass `bucketRole` to that function, in order to get the right creds from the env
- Don't pass `bucketRole` to the class constructor, because we don't care so much about labeling the logger

I can do that - I agree that the logger tag is a nice-to-have rather than a necessity. I guess I'm still confused why passing it to the function is fine, but passing it to the constructor isn't. The "user" you're talking about is the same in both cases (namely, either the GCS target module or the soon-to-exist GCS artifact store module) and in fact the function will be called only about six lines above the constructor, so... 🤔

And I apologize, I'm not meaning to seem like I'm arguing, nor do I mean to belabor the point, but I'd really like to understand the underlying principle here, so I can make good choices going forward.

Reply:
I guess my point is, passing the role (or, basically, an indicator what env vars we want to read) to `getGCSCredsFromEnv` mostly makes sense to me. It's not how I would do this (I'd prefer to have two separate functions that maybe reuse some utility function?), and probably that's why I'm not fully happy with the current implementation, but it doesn't seem wrong.

But passing the role to the constructor just seems... unnecessary? We could theoretically pass lots of different parameters to utility functions to give them more "context" (for loggers and whatnot), but it seems to me like a line we don't want to cross, because it pollutes the API.

Reply:
> Let's also make this a "regular" function :)

I actually meant rewriting arrow function to `function()` function, but moving it outside the class also makes sense to me 👍

Reply:
Ah, okay. Well, in the process of doing the latter I ended up doing the former anyway, so, great! 🙂

Reply:
> I'd prefer to have two separate functions that maybe reuse some utility function

> but it doesn't seem wrong

Okay, so whether or not I change that part, just so I see all the options, you'd do something like 

`getCredsFromEnv = (jsonVarName, pathVarName) => credsObject` 

and then, in the `gcsTarget` code, 

`creds = getCredsFromEnv('CRAFT_GCS_TARGET_CREDENTIALS_JSON', 'CRAFT_GCS_TARGET_CREDENTIALS_PATH')`

?

Reply:
Yes, something like that!

Reply:
Okay, done. See what you think.